New Feature: Load PAT from environment variables

Dragon1573 opened the pull request:
## What's changed

Before querying GitHub REST API for getting release information of `winget.exe`, ...

1. Load `$env:GITHUB_TOKEN` from _Windows Environment Variables_.
2. Check if `$env:GITHUB_TOKEN` begins with `ghp_`, since GitHub REST API uses PAT for authentication.
3. If above 2 steps passed, perform an authorized request.
4. Otherwise, perform unauthorized request as this script used to be.

## Checklist for Pull Requests

- [x] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- Resolve #167997

> [!NOTE]
>
> This pull request is about PowerShell scripts in `Tools/` folder instead of manifests.

---

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/winget-pkgs/pull/168000)

Dragon1573 committed:
New Feature: Load PAT from environment variables

- "$env:GITHUB_TOKEN" should be available.
  - Preffered but not required. If unset, perform unauhorized request as it used to be.
- Easily avoid "API rate limit exeeded" error.

Signed-off-by: Dragon1573 <49941141+Dragon1573@users.noreply.github.com>

Trenly reviewed (changes requested):
You should use parameter splatting so there aren't two calls to `Invoke-RestMethod` that would need to be modified in the future

```powershell
$requestParameters = @{
  Uri='https://api.github.com/repos/microsoft/winget-cli/releases?per_page=100'
}
if ($env:GITHUB_TOKEN && $env:GITHUB_TOKEN.StartsWith('ghp_')) {
    $requestParameters.Add('Authentication','Bearer')
    $requestParameters.Add('Token', $(ConvertTo-SecureString "$env:GITHUB_TOKEN" -AsPlainText))
}
$releasesAPIResponse = Invoke-RestMethod @requestParameters

mdanish-kh commented on Tools/SandboxTest.ps1:
```diff
@@ -83,7 +83,19 @@ New-Item $tempFolder -ItemType Directory -ErrorAction SilentlyContinue | Out-Nul
+  # TODO: Further improvements
+  # Sending authorized requests with GitHub CLI (aka. `gh.exe`) is much more convenient than using `Invoke-RestMethod`.
+  # It would be better integrate this script with GitHub CLI, since developers do not need creating and configuring PAT manually.
+  if ($env:GITHUB_TOKEN && $env:GITHUB_TOKEN.StartsWith('ghp_')) {
```
IIRC only classic tokens start with prefix `ghp_`. Fine-grained tokens seem to begin with `github_pat_`. I would prefer if we don't make an assumption on prefix and make a request first to `https://api.github.com/octocat` to first check if the credentials are valid

https://docs.github.com/en/rest/authentication/authenticating-to-the-rest-api?apiVersion=2022-11-28

Dragon1573 committed:
Apply suggesstions from code review

- Use "parameter splatting" so there aren't 2 calls to `Invoke-RestMethod`
- Suggested by @Trenly

Co-authored-by: Trenly <trenlymc@gmail.com>
Signed-off-by: Dragon1573 <49941141+Dragon1573@users.noreply.github.com>

Dragon1573 committed:
Apply suggestions from code review

- Query a "test endpoint" to see if credential is valid.
- Suggested by @mdanish-kh

Co-authored-by: Muhammad Danish <88161975+mdanish-kh@users.noreply.github.com>
Signed-off-by: Dragon1573 <49941141+Dragon1573@users.noreply.github.com>

Trenly commented on Tools/SandboxTest.ps1:
```diff
@@ -83,7 +83,35 @@ New-Item $tempFolder -ItemType Directory -ErrorAction SilentlyContinue | Out-Nul
+  $requestParameters = @{
+    Uri = 'https://api.github.com/repos/microsoft/winget-cli/releases?per_page=100'
+  }
+  if ($isValidCredential) {
```
Your variable seems backwards. Based on the name, it would  be assumed that `true` means the credential *is* valid. Suggest either renaming the variable or switching the logic

Trenly commented on Tools/SandboxTest.ps1:
```diff
@@ -83,7 +83,35 @@ New-Item $tempFolder -ItemType Directory -ErrorAction SilentlyContinue | Out-Nul
+  if ($env:GITHUB_TOKEN) {
+    $requestParameters.Add('Authentication', 'Bearer')
+    $requestParameters.Add('Token', $(ConvertTo-SecureString "$env:GITHUB_TOKEN" -AsPlainText))
+    $isValidCredential = ([Int16](Invoke-WebRequest @requestParameters).Headers["X-RateLimit-Limit"][0]) -eq 60
```
This has the potential to throw an exception if that specific header isn’t presnt or can’t be converted to an integer. Either check that the header is present first, or use a different parameter like the HTTP response code (if that changes based on whether or not the user is authenticated)

Dragon1573 replied:
API Endpoint `https://api.github.com/octocat` is able to query without authentication. Users can get an ASCII draw of GitHub Octocat. Both (un)authorized request will get `HTTP 200 OK` from this API.

Maybe switching to [REST API endpoints for rate limits](https://docs.github.com/en/rest/rate-limit/rate-limit?apiVersion=2022-11-28) would be better? Response is in JSON format so I can `ConvertFrom-Json` its `.Content` and get the `[Int64]` type `$.resources.core.limit`.

Trenly commented on Tools/SandboxTest.ps1:
```diff
@@ -83,7 +83,35 @@ New-Item $tempFolder -ItemType Directory -ErrorAction SilentlyContinue | Out-Nul
-  $releasesAPIResponse = Invoke-RestMethod 'https://api.github.com/repos/microsoft/winget-cli/releases?per_page=100'
+  # TODO: Further improvements
+  # Sending authorized requests with GitHub CLI (aka. `gh.exe`) is much more convenient than using `Invoke-RestMethod`.
+  # It would be better integrate this script with GitHub CLI, since developers do not need creating and configuring PAT manually.
```
You can check if gh.exe is available. Examples of how to check if a command is present are in YamlCreate - since it checks for winget.exe in a few instances.

Dragon1573 replied:
Check `gh.exe` with `Get-Command`. If it's available, always prefer using its _GitHub OAuth Token_ (begin with `gho_`) as `$env:GITHUB_TOKEN` for further queries.

Dragon1573 committed:
Apply suggestions from code review

- Switch to API endpoint specific for checking rate limits.

Co-authored-by: Trenly <trenlymc@gmail.com>
Signed-off-by: Dragon1573 <49941141+Dragon1573@users.noreply.github.com>

Trenly commented on Tools/SandboxTest.ps1:
```diff
@@ -90,12 +90,12 @@ function Get-Release {
     $requestParameters.Add('Authentication', 'Bearer')
     $requestParameters.Add('Token', $(ConvertTo-SecureString "$env:GITHUB_TOKEN" -AsPlainText))
-    $isValidCredential = ([Int16](Invoke-WebRequest @requestParameters).Headers["X-RateLimit-Limit"][0]) -eq 60
+    $isValidCredential = (ConvertFrom-Json (Invoke-WebRequest @requestParameters).Headers["X-RateLimit-Limit"]).resources.core.limit -ne 60
```
Instead of using `ConvertFrom-Json` with `Invoke-WebRequest` you can use `Invoke-RestMethod` to perform the JSON conversion automatically

Dragon1573 committed:
Apply suggestions from code review

- Check if GitHub CLI is available.
- Always prefer GitHub OAuth Token provided by GitHub CLI.

Signed-off-by: Dragon1573 <49941141+Dragon1573@users.noreply.github.com>

Trenly reviewed (commented):
I think using separate functions for validating the token would be best, since then if additional API calls are added in the future we can call that function to see if it is valid.

I'll rase a PR into your branch so that what I'm thinking gets applied directly - will likely be faster that way

Trenly commented:
I wasn't able to make a PR directly into your branch due to merge conflicts, so I raised it into `main` in your fork. I can raise a copy of the PR here, which is probably the easiest

Dragon1573 commented:
"Allow edits and access to secrets by maintainers" is now enabled for this pull request ❤️

Trenly commented:
> "Allow edits and access to secrets by maintainers" is now enabled for this pull request ❤️

I'm not a maintainer though 😄

Dragon1573 committed:
Apply suggestions from code review

- If there exists `gh.exe`, just query release info with it.
- If `$env:GITHUB_TOKEN` is not present, just skip checking rate limits.

Co-authored-by: Trenly <trenlymc@gmail.com>
Signed-off-by: Dragon1573 <49941141+Dragon1573@users.noreply.github.com>

Dragon1573 commented:
Since you created #168384, shall I close this (#168000) without being merged?

AsciiWolf commented on Tools/SandboxTest.ps1:
```diff
@@ -83,7 +83,44 @@ New-Item $tempFolder -ItemType Directory -ErrorAction SilentlyContinue | Out-Nul
+  else {
+    <# Use original Powershell `Invoke-WebRequest` and `Invoke-RestMethod` without authentication. #>
+
+    Write-Warning 'WARNING: $env.GITHUB_TOKEN not set! You may encounter "API rate limit exeeded" error!'
```
```suggestion
    Write-Warning 'WARNING: $env.GITHUB_TOKEN not set! You may encounter "API rate limit exceeded" error!'
```

Small typo.

Dragon1573 committed:
Apply suggestions from code review

- Fix typo error.

Co-authored-by: AsciiWolf <mail@asciiwolf.com>

Dragon1573 commented:
Close with reason: https://github.com/microsoft/winget-pkgs/pull/168000#issuecomment-2299291512
